package.json

```diff
@@ -12,24 +12,32 @@
     "test": "jest"
   },
   "dependencies": {
+    "@aws-sdk/client-s3": "^3.53.1",
+    "@aws-sdk/lib-storage": "^3.53.1",
     "luxon": "^2.3.1",
-    "superstruct": "^0.15.3"
+    "superstruct": "^0.15.3",
+    "tar": "^6.1.11"
   },
   "devDependencies": {
     "@tsconfig/node14": "^1.0.1",
     "@types/aws-lambda": "^8.10.92",
     "@types/jest": "^27.4.0",
     "@types/luxon": "^2.0.9",
     "@types/node": "^17.0.17",
+    "@types/s3rver": "^3.7.0",
+    "@types/tar": "^6.1.1",
     "@types/uuid": "^8.3.4",
     "@typescript-eslint/eslint-plugin": "^5.11.0",
     "@typescript-eslint/parser": "^5.11.0",
     "esbuild": "^0.14.21",
     "eslint": "^8.8.0",
     "eslint-config-prettier": "^8.3.0",
     "fishery": "^2.1.0",
+    "get-stream": "^6.0.1",
     "jest": "^27.5.1",
     "prettier": "^2.5.1",
+    "readdirp": "^3.6.0",
+    "s3rver": "^3.7.1",
```

Comment:
> I've tried to trace out its dependencies, and those dependencies' dependencies, and so on, and I weep for the (certain) dependabot churn it will bring us

Aren't deps-of-deps irrelevant for dependabot purposes, except in the case of security patches? We'd only be prompted to update if `s3rver` itself released a new version, which only seems to happen ~twice a year — much _less_ often than `aws-sdk-client-mock`, actually, which had 12 updates last year. Not to invalidate the other concerns you mention.

> I'm only picking on `s3rver` here because the description was too honest and made it sound like we _could_ have good, if not perfect, test coverage even without it, using dependencies we already have.

I'm not sure what you mean here — `aws-sdk-client-mock` is not a dependency we already have.

> So I guess, can you help me understand just how big the benefit of `s3rver` is here?

I started out writing these tests using `aws-sdk-client-mock`, and eventually switched to `s3rver` due to some issues that were bogging me down (I didn't document what they were, unfortunately, since I assumed `s3rver` would be an uncontroversial choice). The benefits as I see them:

* Tests can treat the code as a black box, only verifying that correct inputs give correct outputs, without needing knowledge of the internal implementation ("when X,Y,Z are in an S3 bucket, and I run the code, I should see output ABC in the S3 bucket"). With client mocking, the tests need to know that the code is going to use e.g. a `ListObjectsV2Command`, and mock that specific command. This applies when it's not even _our_ code that's using the command, like when using the `paginateListObjectsV2` helper. To analogize a bit, this feels like testing code that uses Ecto by mocking the return values of specific SQL statements, rather than setting up test data in a real database. (We certainly can't always do things like this, but where it's easily possible, to me it has a clear value in keeping tests uncoupled from the implementation.)
* With client mocking, it's possible for tests to accidentally set up responses in a way that doesn't match what the real S3 would return in that scenario. `aws-sdk-client-mock` does guard against this on the type level, but it can't do so on a contextual level. For example, the type of an attribute could be `undefined | string`, meaning you'd always be _allowed_ to mock it either way, but in reality it might only be undefined given certain conditions (like "the bucket has no objects") that can't be expressed in the type system. I can't point to any specific examples here since it's to do with avoiding pitfalls that by definition we wouldn't know about — though I don't think there are really a _lot_ of these within the scope of what we're doing, so weigh that accordingly.
* `s3rver` has been considered "stable" (now on version 3.x) for several years, and has very infrequent releases. `aws-sdk-client-mock` is very new and still considered "unstable" (currently on 0.x and has gone through 6 "major" versions in the 1 year it has existed). So if you put some value on that, the former seems like the safer choice. (I'm refraining from saying myself that this is valuable since then I'd be a real hypocrite for choosing the 0.x `superstruct` over the 3.x `zod`, though in that case both have zero dependencies and are small libraries to begin with, so maybe the calculus is different)

So, there's some things to consider. While of course I prefer the current approach, and wouldn't exactly look forward to rewriting the tests to use client mocking (this has been a rather long project already), I think it's also a reasonable option and would take a second crack at it if you think the reduced dependency footprint justifies the effort.

Reply:
I see. Thanks for this.

I had gotten a bit snowblind scrolling through `package-lock.json` and the dozens upon dozens of `@aws-sdk/*` packages, which are all part of the `aws` GitHub account, and missed that `aws-sdk-client-mock` is _not_ one of those, and in fact, not official at all. (Sorry, I should have followed the link in your PR description.) While its transitive dependencies are fewer than `s3rver`, it's still a pretty hefty graph.

Given that, and given your examples here of how `s3rver` is better, I'm definitely 👍 about using `s3rver`.

Reply:
(The endorsement does have other benefits to consider like the potential for other eyes on the project, and its chances of being abandoned, but I'm okay with `s3rver` on that front given its history.)